src/game/ui/tdroplist.h

```diff
@@ -0,0 +1,301 @@
+/**
+ * @file
+ *
+ * @author OmniBlade
+ * @author CCHyper
+ * @author tomsons26
+ *
+ * @brief Class for representing a droppable lists of selectable options.
+ *
+ * @copyright Chronoshift is free software: you can redistribute it and/or
+ *            modify it under the terms of the GNU General Public License
+ *            as published by the Free Software Foundation, either version
+ *            2 of the License, or (at your option) any later version.
+ *            A full copy of the GNU General Public License can be found in
+ *            LICENSE
+ */
+#pragma once
+
+#ifndef TDROPLIST_H
+#define TDROPLIST_H
+
+#include "always.h"
+#include "edit.h"
+#include "tlist.h"
+#include "shapebtn.h"
+#include "dialog.h"
+#include "drawshape.h"
+
+template<typename T>
+class TDropListClass : public EditClass
+{
+public:
+    TDropListClass(int id, char *text_buff, int text_size, TextPrintType style, int x, int y, int w, int h,
+        void *up_btn_shape, void *down_btn_shape);
+    TDropListClass(TDropListClass<T> &that);
+    virtual ~TDropListClass() {}
+
+    virtual TDropListClass<T> &Add(LinkClass &link) override;
+    virtual TDropListClass<T> &Add_Tail(LinkClass &link) override;
+    virtual TDropListClass<T> &Add_Head(LinkClass &link) override;
+    virtual void Zap() override;
+    virtual TDropListClass<T> *Remove() override;
+    virtual void Flag_To_Redraw() override;
+    virtual void Peer_To_Peer(unsigned flags, KeyNumType &key, ControlClass &peer) override;
+    virtual void Clear_Focus() override;
+    virtual void Set_Position(int x, int y) override;
+    virtual int Add_Item(T item);
+    virtual T Current_Item() const;
+    virtual int Current_Index();
+    virtual void Set_Selected_Index(int index);
+    virtual void Set_Selected_Index(T item);
+    virtual int Count() const;
+    virtual T Get_Item(int const string_index) const;
+
+    TDropListClass &operator=(TDropListClass &that);
+
+    void Expand();
+    void Collapse();
+
+protected:
+#ifndef CHRONOSHIFT_NO_BITFIELDS
+    BOOL IsExpanded : 1; // & 1
+#else
+    bool IsExpanded;
+#endif
+    int DropHeight;
+    ShapeButtonClass DropButton;
+    TListClass<T> DropList;
+};
+
+template<typename T>
+inline TDropListClass<T> &TDropListClass<T>::operator=(TDropListClass<T> &that)
+{
+    if (this != &that) {
+        EditClass::operator=(that);
+        IsExpanded = that.IsExpanded;
+        DropHeight = that.DropHeight;
+        DropButton = that.DropButton;
+        DropList = that.DropList;
+        DropButton.Make_Peer(*this);
+        DropList.Make_Peer(*this);
+    }
+
+    return *this;
+}
+
+//same as DropListClass
```

Comment:
Don't put comments like this, put proper comment blocks and add an` @info Method implementation is the same as used in DropListClass` or something similar though that isn't all that useful to someone looking to use the classes API, only to someone investigating its internals.